src/test-util.ts

```diff
@@ -32,3 +34,195 @@ export function b(
   }
   return utf8.encode(s);
 }
+
+export function defineTestFunctions(self: Window & typeof globalThis): void {
+  // Mocha functions aren't available when importing from a new page.
+  if (typeof self.suiteSetup !== 'function') {
+    // eslint-disable-next-line @typescript-eslint/no-empty-function
+    const noop = function () {};
+
+    self.setup = noop;
+    self.teardown = noop;
+    self.suiteSetup = noop;
+    self.suiteTeardown = noop;
+
+    // eslint-disable-next-line @typescript-eslint/no-empty-function
+    const suite = function () {};
+    suite.only = noop;
+    suite.skip = noop;
+    self.suite = suite as unknown as Mocha.SuiteFunction;
+
+    // eslint-disable-next-line @typescript-eslint/no-empty-function
+    const test = function () {};
+    test.only = noop;
+    test.skip = noop;
+    test.retries = noop;
+    self.test = test as unknown as Mocha.TestFunction;
+  }
+}
+
+defineTestFunctions(self);
+
+type Response = {
+  tabId: string;
+  id: string;
+  result?: unknown;
+  error?: unknown;
+};
+
+type LogResponse = {
+  level: string;
+  message: string;
+};
+
+export async function initializeNewTab(event: MessageEvent): Promise<void> {
+  const {tabId, path} = event.data;
+
+  ['error', 'warn', 'info', 'debug', 'log'].forEach(level => {
+    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+    // @ts-ignore No index signature of type 'string' was found on 'Console'.
+    console[level] = (...data: unknown[]) => {
+      window.opener.postMessage(
+        {tabId, id: 'log', result: {level, message: data.join(' ')}},
```

Comment:
just `message: data` seems more accurate (and flexible)

Reply:
Yep, that's right. I had struggled to figure out how to then pass the resulting array back to message.log() and friends, but just got `...data` working. Thanks!